larpandoracontent/LArThreeDReco/LArCosmicRay/AmbiguousDeltaRayTool.cc

```diff
@@ -0,0 +1,130 @@
+/**
+ *  @file   larpandoracontent/LArThreeDReco/LArCosmicRay/AmbiguousDeltaRayTool.cc
+ *
+ *  @brief  Implementation of the ambiguous delta ray tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArThreeDReco/LArCosmicRay/AmbiguousDeltaRayTool.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+AmbiguousDeltaRayTool::AmbiguousDeltaRayTool() :
+    m_maxGoodMatchReducedChiSquared(1.f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool AmbiguousDeltaRayTool::Run(ThreeViewDeltaRayMatchingAlgorithm *const pAlgorithm, TensorType &overlapTensor)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+       std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    this->ExamineConnectedElements(pAlgorithm, overlapTensor);
+
+    // ATTN: Prevent tensor tool loop running again
+    return false;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void AmbiguousDeltaRayTool::ExamineConnectedElements(ThreeViewDeltaRayMatchingAlgorithm *const pAlgorithm, TensorType &overlapTensor) const
+{
+    ClusterVector sortedKeyClusters;
+    overlapTensor.GetSortedKeyClusters(sortedKeyClusters);
+
+    ClusterSet usedClusters;
+    ClusterSet usedKeyClusters;
+    ProtoParticleVector protoParticleVector;
+
+    for (const Cluster *const pKeyCluster : sortedKeyClusters)
+    {
+        if (usedKeyClusters.count(pKeyCluster))
+            continue;
+
+        TensorType::ElementList elementList;
+        overlapTensor.GetConnectedElements(pKeyCluster, true, elementList);
+
+        for (const TensorType::Element &element : elementList)
+            usedKeyClusters.insert(element.GetClusterU());
+
+        if (elementList.size() < 2)
+            continue;
+
+        this->PickOutGoodMatches(elementList, usedClusters, protoParticleVector);
+    }
+
+    if (!protoParticleVector.empty())
+        pAlgorithm->CreatePfos(protoParticleVector);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void AmbiguousDeltaRayTool::PickOutGoodMatches(const TensorType::ElementList &elementList, ClusterSet &usedClusters, ProtoParticleVector &protoParticleVector) const
+{
+    bool found(true);
+    
+    while (found)
```

Comment:
Prefer do {...} while () when you know the code in the loop should be executed at least once